src/tree/hist/evaluate_splits.h

```diff
@@ -156,36 +188,134 @@ template <typename GradientSumT, typename ExpandEntry> class HistEvaluator {
         tloc_candidates[i * n_threads_ + j] = entries[i];
       }
     }
+    std::vector<ExpandEntry> tloc_candidates_sub;
+    if (entries_sub.size() != 0) {
+      tloc_candidates_sub.resize(omp_get_max_threads() * entries_sub.size());
+      for (size_t i = 0; i < entries_sub.size(); ++i) {
+        for (decltype(n_threads_) j = 0; j < n_threads_; ++j) {
+          tloc_candidates_sub[i * n_threads_ + j] = entries_sub[i];
+        }
+      }
+    }
     auto evaluator = tree_evaluator_.GetEvaluator();
 
-    common::ParallelFor2d(space, n_threads_, [&](size_t nidx_in_set, common::Range1d r) {
-      auto tidx = omp_get_thread_num();
-      auto entry = &tloc_candidates[n_threads_ * nidx_in_set + tidx];
+    const size_t num_blocks_in_space = space.Size();
+    const bool is_dense_and_root = gidx.IsDense() && entries_sub.size() == 0;
+    #pragma omp parallel for schedule(guided)
+    for (bst_omp_uint task_id = 0; task_id < num_blocks_in_space; ++task_id) {
+      const auto tidx = omp_get_thread_num();
+      size_t nidx_in_set = space.GetFirstDimension(task_id);
+      common::Range1d r = space.GetRange(task_id);
+
+      ExpandEntry* entry = &tloc_candidates[n_threads_ * nidx_in_set + tidx];
       auto best = &entry->split;
-      auto nidx = entry->nid;
-      auto histogram = hist[nidx];
+      int nidx = entry->nid;
+      typename common::HistCollection<GradientSumT>::GHistRowT histogram = hist[nidx];
       auto features_set = features[nidx_in_set]->ConstHostSpan();
+
+      GradientSumT* parent_hist = nullptr;
+      GradientSumT* largest_hist = nullptr;
+      ExpandEntry* entry_s = nullptr;
+      int nidx_s = 0;
+      typename common::HistCollection<GradientSumT>::GHistRowT histogram_s;
+      if (entries_sub.size() != 0) {
+        entry_s = &tloc_candidates_sub[n_threads_ * nidx_in_set + tidx];
+        nidx_s = entry_s->nid;
+        histogram_s = hist[nidx_s];
+        const size_t parent_id = (*p_tree)[nidx_s].Parent();
+        parent_hist = reinterpret_cast<GradientSumT*>(hist[parent_id].data());
+        largest_hist = reinterpret_cast<GradientSumT*>(histogram_s.data());
+      }
+      size_t begin = 2*cut_ptr[features_set[r.begin()]];
+      size_t end = 2*cut_ptr[features_set[r.end()-1] + 1];
+
+      GradientSumT* dest_hist = reinterpret_cast<GradientSumT*>(histogram.data());
+      if (p_opt_partition_builder->threads_id_for_nodes[nidx].size() != 0
+          && !is_distributed_ && !is_dense_and_root && !colsample_enabled) {
+        const size_t first_thread_id = p_opt_partition_builder->threads_id_for_nodes[nidx][0];
+        std::vector<uint16_t>& local_thread_mapping = local_threads_mapping[first_thread_id];
```

Comment:
Again, be aware of memory usage for wide datasets.

Reply:
By this mapping I fixed the problem for increased memory consumption in first commit of this PR.
Now there is no big memory usage increase: for 'lossguided' mode memory usage is similar, for 'depthwise' computational mode it's not big increase (epsilon 20% only, airline 3%).